Add config loader and examples

roycaihw opened the pull request:
Prototyping generated Go client. This PR has three commits:

1. `./examples` contains two basic examples tested with the client and config packages (out-of-cluster example tested with local cluster and kube-up (gcp). tls and auth work). 

2. `./kubernetes/config/api` contains type definition and deepcopy utility copied from `k8s.io/client-go` that are used for parsing and persist kube config yaml file.

3. `./kubernetes/config` contains in-cluster and out-of-cluster config loaders with TLS and several auth support.
     **NOTE:** The config loader should be split into `go-base` repo. Have it here just for simple test.

roycaihw commented:
cc @yliaog @mbohlool

roycaihw commented:
@yliaog I added unit tests to test loading authentication and TLS. Also manually tested refreshing GCP token and persisting kubeconfig file end-to-end. Please take a look

yliaog commented on kubernetes/config/api/types.go:
```diff
@@ -0,0 +1,204 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
```
could you add a README to document how they are generated, what they are used for?

roycaihw replied:
updated

yliaog commented:
you might want to change the PR description as it still says 'two commits'

roycaihw commented:
> you might want to change the PR description as it still says 'two commits'

@yliaog updated

yliaog commented on kubernetes/config/gcp.go:
```diff
@@ -0,0 +1,111 @@
+	setUserWithName(l.rawConfig.AuthInfos, l.currentContext.AuthInfo, &l.user)
+	// Persist kube config file
+	if !l.skipConfigPersist {
+		err = l.persistConfig()
```
why not using err :=

roycaihw replied:
I used to avoid shadowing error in golang if there is no need for multi-scope error return. Is it good practice to always shadow error in if statement, if error is the only return value from `foo()`? Like:

```
if err := foo(); err != nil {
    return bar(err)
}
```

instead of

```
if err = foo(); err != nil {
    return bar(err)
}
```

I agree using shadowing is more readable. Thanks for pointing out!

yliaog commented on kubernetes/config/util.go:
```diff
@@ -0,0 +1,111 @@
+
+	for i, u := range users {
+		if u.Name == name {
+			if userFound == true {
```
could be just: if userFound { }

yliaog commented on kubernetes/config/util.go:
```diff
@@ -0,0 +1,111 @@
+			userFound = true
+		}
+	}
+	if userFound == false {
```
could be just: if !userFound {

yliaog commented on kubernetes/config/incluster_config.go:
```diff
@@ -0,0 +1,74 @@
+		Host:          net.JoinHostPort(host, port),
+		Scheme:        "https",
+		DefaultHeader: map[string]string{"Authentication": "Bearer " + string(token)},
+		UserAgent:     "Swagger-Codegen/0.1.0a1/go",
```
not sure whether this should be updated

yliaog replied:
also it should be a const

yliaog commented on kubernetes/config/kube_config.go:
```diff
@@ -0,0 +1,281 @@
+)
+
+const (
+	defaultUserAgent          = "Swagger-Codegen/0.1.0a1/go"
```
again, probably it should be updated.

roycaihw replied:
Yes I think it should be generated. Probably in a release automation as a future step

yliaog commented on kubernetes/config/kube_config.go:
```diff
@@ -0,0 +1,281 @@
+	gcLoader GoogleCredentialLoader
+}
+
+// RestConfig contains the information that a rest client needed to talk with a server
```
s/needed/needs/

yliaog commented on kubernetes/config/kube_config.go:
```diff
@@ -0,0 +1,281 @@
+	l.loadAuthentication()
+
+	err := l.loadClusterInfo()
+	if err != nil {
```
i think Go recommends to combine the above two lines into one

yliaog commented on kubernetes/config/kube_config.go:
```diff
@@ -0,0 +1,281 @@
+	l.restConfig.basePath = strings.TrimRight(l.cluster.Server, "/")
+
+	u, err := url.Parse(l.cluster.Server)
+	if err != nil {
```
combine the two lines?

roycaihw replied:
For this case `u` would be scoped in the if statement. And golint complain if I do "if...else..." block with `return` inside. I don't see much gain in readability in this case

yliaog replied:
sounds reasonable.

yliaog commented on kubernetes/config/kube_config.go:
```diff
@@ -0,0 +1,281 @@
+
+	// Persist kube config file
+	if !l.skipConfigPersist {
+		err = l.persistConfig()
```
err :=

roycaihw commented:
@yliaog update. PTAL

yliaog commented on kubernetes/config/util_test.go:
```diff
@@ -0,0 +1,59 @@
+)
+
+func TestSetUserWithName(t *testing.T) {
+	tcs := []struct {
```
fine to merge, but better to add tests for other functions

yliaog commented on kubernetes/config/kube_config_test.go:
```diff
@@ -0,0 +1,459 @@
+		t.Errorf("context %v, unexpected error setting up fake config: %v", tc.ActiveContext, err)
+	}
+
+	err = ioutil.WriteFile("/tmp/client-ssl-test-cert-file", testCertAuthBase64, 0644)
```
better to use temp file, like http://golang-examples.tumblr.com/post/81467564564/create-a-temporary-file, 1. make it unique, 2. clean it up

yliaog commented on examples/in-cluster-client-configuration/main.go:
```diff
@@ -0,0 +1,46 @@
+	"fmt"
+	"time"
+
+	"k8s.io/go/kubernetes/client"
```
k8s.io/client/... is probably better, as this is golang's import, there is no need to say 'go' anymore.

roycaihw commented:
@yliaog updated

yliaog commented:
it looks good to me, could you squash the commits?

roycaihw committed:
Add examples for config loader and basic client

roycaihw committed:
Migrate config Go type and deepcopy from k8s.io/client-go

roycaihw committed:
Add basic kube config loader

NOTE: The config loader should be split into go-base repo. Have it
      here just for simple test.

roycaihw commented:
@yliaog Squashed. Thanks for reviewing!

yliaog commented:
/lgtm
